text/0118-priming-hints.md

```diff
@@ -0,0 +1,174 @@
+
+In order to increase the scope of priming possible, we suggest allowing component developers to provide priming hints. Priming hints can guide priming mechanisms into artificially substituting in property values that introduce more component sub-trees into the component's markup. This would allow the priming mechanism to override property value calculations and mimick happy path tree expansion, introducing a deeper component sub-tree to prime.
+
+## Example
+
+Consider a "bar" container that acts like a tab set. It holds a collection of items, and allows the user to interactively choose an item to be displayed in its default slot:
+
+```html
+<!-- bar.html -->
+<template>
+    <div class={computedClass} title={title}>
+        <bar-header></bar-header>
+        <slot></slot>
+    </div>
+</template>
+```
+
+The items look like the following:
+
+```js
+export default class BarItem extends LightningElement {
+    @track _loadContent = false;
+
+    @api
+    loadContent() {
+        this._loadContent = true;
+
+        this.dispatchEvent(new CustomEvent('active'));
+    }
+
+    ...
+}
+```
+
+```html
+<!-- barItem.html -->
+<template>
+    <template if:true={_loadContent}>
+        <slot></slot>
+    </template>
+</template>
+```
+
+The container lazily loads the contents of its children only when they are first selected for view. The bar container can control the currently visible barItem by calling the loadContent() method to load its content and using CSS to hide any previously displayed barItem.
+
+For a prefetch mechanism based on static analysis, this structure poses a barrier. The conditional loading of barItem is governed by a member _loadContent that is triggered by user interaction. If the prefetch mechanism extracts composition information from markup, and executes resolution based on that analysis without a DOM or user interaction, the best it could hope for it priming the contents of the barItem that is first displayed by default. For some use cases of priming that is reasonable (such as for speeding up render while online). However, for priming for an offline use case one may desire a greater scope of priming - priming the contents of all the barItems. In this case, it would be desirable to provide a hint to the priming mechanism to say "please assume this if:true condition is true for priming purposes".
+
+## Design
+
+We propose adding new meaningful comments to LWC which allow components to opt-in to marking properties that a priming mechanisms could assume to be true. This will allow priming mechanisms to prime a greater scope of component contents.
+
+These comments would be used on property getters, and their format would be:
+
+```js
+/**
+ * priming-hint: <scope> <value>
```

Comment:
You left out some details for `value`; my initial thought was that `value` should just default to a boolean `true`. We could expand upon priming hints to support other values if necessary, but for the initial draft it could just be a boolean `true`. Thoughts?

If you disagree, we should discuss what valid values could be provided; I would suggest only simple literals; e.g., strings, numbers, and booleans to minimize having to parse the comment more than necessary.

Reply:
I agree. this is just leaving the door for non-boolean values in the future.